src/celeritas/io/NeutronXsReader.cc

```diff
@@ -0,0 +1,99 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/io/NeutronXsReader.cc
+//---------------------------------------------------------------------------//
+#include "NeutronXsReader.hh"
+
+#include <fstream>
+#include <vector>
+
+#include "corecel/Assert.hh"
+#include "corecel/Types.hh"
+#include "corecel/io/Logger.hh"
+#include "corecel/math/Algorithms.hh"
+#include "corecel/sys/Environment.hh"
+#include "celeritas/Units.hh"
+#include "celeritas/io/ImportLivermorePE.hh"
+#include "celeritas/io/ImportPhysicsVector.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct the reader using the G4PARTICLEXSDATA environment variable to get
+ * the path to the data.
+ */
+NeutronXsReader::NeutronXsReader()
+{
+    std::string const& dir = celeritas::getenv("G4PARTICLEXSDATA");
+    CELER_VALIDATE(!dir.empty(),
+                   << "environment variable G4PARTICLEXSDATA is not defined"
+                      " (needed to locate Livermore data)");
+    path_ = dir + "/neutron";
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct the reader with the path to the directory containing the data.
+ */
+NeutronXsReader::NeutronXsReader(char const* path) : path_(path)
+{
+    CELER_EXPECT(!path_.empty());
+    if (path_.back() == '/')
+    {
+        path_.pop_back();
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Read the data for the given element.
+ */
+NeutronXsReader::result_type
+NeutronXsReader::operator()(AtomicNumber atomic_number) const
+{
+    CELER_EXPECT(atomic_number && atomic_number < AtomicNumber{92});
+
+    std::string z_str = std::to_string(atomic_number.unchecked_get());
+    CELER_LOG(debug) << "Reading neutron elastic xs data for Z=" << z_str;
+
+    result_type result;
+
+    // Read neutron elastic cross section data for the given atomic_number
+    {
+        std::string filename = path_ + "/el" + z_str;
+        std::ifstream infile;
+        infile.open(filename);
+        CELER_VALIDATE(infile.is_open(),
```

Comment:
```suggestion
        std::ifstream infile(filename);
        CELER_VALIDATE(infile,
```
I think `operator bool` for istream catches more error conditions than `is_open`: https://stackoverflow.com/a/14920476/118160

Reply:
Okay, I thought that using file open is more thread safe, but my recollection is hazy in these days. I take your suggestion.

Reply:
I've never heard that, but I've also never paid any attention to thread safety before working with Geant4 😅 I think the two should be essentially equivalent.